File: src/main/java/net/omori_sunny/create_waterparked/ponder/WaterslidePonderScenes.java
package net.omori_sunny.create_waterparked.ponder;

import net.createmod.ponder.api.registration.PonderSceneRegistrationHelper;
import net.minecraft.core.registries.BuiltInRegistries;
import net.minecraft.resources.ResourceLocation;
import net.omori_sunny.create_waterparked.content.attachment.ModSlideAttachments;
import net.omori_sunny.create_waterparked.content.attachment.SlideAttachmentType;
import net.omori_sunny.create_waterparked.content.attachment.SlideAttachmentTypes;
import net.omori_sunny.create_waterparked.content.registry.ModItems;
import org.jetbrains.annotations.Nullable;

import java.util.Map;

public class WaterslidePonderScenes {

    public static final String CONNECT_SCHEMATIC = "waterslide_anchor/ponder_connect";
    public static final String SLIDE_USE_SCHEMATIC = "waterslide_anchor/slide_ponder_0";
    public static final String SLIDE_SECTOR_SCHEMATIC = "waterslide_anchor/slide_ponder_1";
    public static final String SECTOR_SCENE_ID = "slide_ponder_1";
    public static final String GHOST_SCENE_ID = "slide_ponder_2";
    public static final String ATTACHMENT_SCHEMATIC = "slide_attachment/sa_ponder_0";
    public static final String ATTACHMENT_SCENE_ID = "sa_ponder_0";
    public static final String DOOR_SCENE_ID = "door_ponder_0";

    private static final Map<ResourceLocation, ResourceLocation> SCHEMATIC_PATHS = Map.of(
        ResourceLocation.fromNamespaceAndPath("create_waterparked", "ponder_connect"),
        ResourceLocation.fromNamespaceAndPath("create_waterparked", CONNECT_SCHEMATIC),
        ResourceLocation.fromNamespaceAndPath("create_waterparked", "slide_ponder_0"),
        ResourceLocation.fromNamespaceAndPath("create_waterparked", SLIDE_USE_SCHEMATIC),
        ResourceLocation.fromNamespaceAndPath("create_waterparked", SECTOR_SCENE_ID),
        ResourceLocation.fromNamespaceAndPath("create_waterparked", SLIDE_SECTOR_SCHEMATIC),
        ResourceLocation.fromNamespaceAndPath("create_waterparked", GHOST_SCENE_ID),
        ResourceLocation.fromNamespaceAndPath("create_waterparked", SLIDE_SECTOR_SCHEMATIC),
        ResourceLocation.fromNamespaceAndPath("create_waterparked", ATTACHMENT_SCENE_ID),
        ResourceLocation.fromNamespaceAndPath("create_waterparked", ATTACHMENT_SCHEMATIC),
        ResourceLocation.fromNamespaceAndPath("create_waterparked", DOOR_SCENE_ID),
        ResourceLocation.fromNamespaceAndPath("create_waterparked", ATTACHMENT_SCHEMATIC)
    );

    private WaterslidePonderScenes() {
    }

    @Nullable
    public static ResourceLocation schematicPathFor(ResourceLocation sceneId) {
        return SCHEMATIC_PATHS.get(sceneId);
    }

    public static void register(PonderSceneRegistrationHelper<ResourceLocation> registry) {
        ResourceLocation anchor = BuiltInRegistries.ITEM.getKey(ModItems.INSTANCE.getWATERSLIDE_ANCHOR());
        ResourceLocation track = BuiltInRegistries.ITEM.getKey(ModItems.INSTANCE.getWATERSLIDE_TRACK());
        registry.addStoryBoard(anchor, SLIDE_USE_SCHEMATIC, WaterslidePonderScene::useSlideTrack, new ResourceLocation[0]);
        registry.addStoryBoard(track, SLIDE_USE_SCHEMATIC, WaterslidePonderScene::useSlideTrack, new ResourceLocation[0]);
        registry.addStoryBoard(anchor, SLIDE_SECTOR_SCHEMATIC, WaterslidePonderScene::sectorSystem, new ResourceLocation[0]);
        registry.addStoryBoard(track, SLIDE_SECTOR_SCHEMATIC, WaterslidePonderScene::sectorSystem, new ResourceLocation[0]);
        registry.addStoryBoard(anchor, SLIDE_SECTOR_SCHEMATIC, WaterslidePonderScene::ghostBlocks, new ResourceLocation[0]);
        registry.addStoryBoard(track, SLIDE_SECTOR_SCHEMATIC, WaterslidePonderScene::ghostBlocks, new ResourceLocation[0]);
        for (SlideAttachmentType type : SlideAttachmentTypes.INSTANCE.all()) {
            ResourceLocation binding = BuiltInRegistries.ITEM.getKey(type.getItem()
                .get());
            registry.addStoryBoard(binding, ATTACHMENT_SCHEMATIC, WaterslidePonderScene::placeAttachment, new ResourceLocation[0]);
        }
        ResourceLocation door = BuiltInRegistries.ITEM.getKey(ModSlideAttachments.INSTANCE.getMECHANICAL_DOOR()
            .getItem()
            .get());
        registry.addStoryBoard(door, ATTACHMENT_SCHEMATIC, WaterslidePonderScene::mechanicalDoor, new ResourceLocation[0]);
    }
}

File: src/main/kotlin/net/omori_sunny/create_waterparked/ponder/WaterslidePonderScenes.kt
package net.omori_sunny.create_waterparked.ponder

import net.createmod.ponder.api.registration.PonderSceneRegistrationHelper
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.resources.ResourceLocation
import net.omori_sunny.create_waterparked.content.attachment.ModSlideAttachments
import net.omori_sunny.create_waterparked.content.attachment.SlideAttachmentTypes
import net.omori_sunny.create_waterparked.content.registry.ModItems

object WaterslidePonderScenes {

    const val CONNECT_SCHEMATIC = "waterslide_anchor/ponder_connect"
    const val SLIDE_USE_SCHEMATIC = "waterslide_anchor/slide_ponder_0"
    const val SLIDE_SECTOR_SCHEMATIC = "waterslide_anchor/slide_ponder_1"
    const val SECTOR_SCENE_ID = "slide_ponder_1"
    const val GHOST_SCENE_ID = "slide_ponder_2"
    const val ATTACHMENT_SCHEMATIC = "slide_attachment/sa_ponder_0"
    const val ATTACHMENT_SCENE_ID = "sa_ponder_0"
    const val DOOR_SCENE_ID = "door_ponder_0"
    const val DETECTOR_SCENE_ID = "detector_ponder_0"

    private val SCHEMATIC_PATHS: Map<ResourceLocation, ResourceLocation> = mapOf(
        path("ponder_connect") to path(CONNECT_SCHEMATIC),
        path("slide_ponder_0") to path(SLIDE_USE_SCHEMATIC),
        path(SECTOR_SCENE_ID) to path(SLIDE_SECTOR_SCHEMATIC),
        path(GHOST_SCENE_ID) to path(SLIDE_SECTOR_SCHEMATIC),
        path(ATTACHMENT_SCENE_ID) to path(ATTACHMENT_SCHEMATIC),
        path(DOOR_SCENE_ID) to path(ATTACHMENT_SCHEMATIC),
        path(DETECTOR_SCENE_ID) to path(ATTACHMENT_SCHEMATIC)
    )

    private fun path(id: String): ResourceLocation =
        ResourceLocation.fromNamespaceAndPath("create_waterparked", id)

    @JvmStatic
    fun schematicPathFor(sceneId: ResourceLocation): ResourceLocation? = SCHEMATIC_PATHS[sceneId]

    @JvmStatic
    fun register(registry: PonderSceneRegistrationHelper<ResourceLocation>) {
        val anchor = BuiltInRegistries.ITEM.getKey(ModItems.WATERSLIDE_ANCHOR)
        val track = BuiltInRegistries.ITEM.getKey(ModItems.WATERSLIDE_TRACK)
        registry.addStoryBoard(anchor, SLIDE_USE_SCHEMATIC, WaterslidePonderScene::useSlideTrack)
        registry.addStoryBoard(track, SLIDE_USE_SCHEMATIC, WaterslidePonderScene::useSlideTrack)
        registry.addStoryBoard(anchor, SLIDE_SECTOR_SCHEMATIC, WaterslidePonderScene::sectorSystem)
        registry.addStoryBoard(track, SLIDE_SECTOR_SCHEMATIC, WaterslidePonderScene::sectorSystem)
        registry.addStoryBoard(anchor, SLIDE_SECTOR_SCHEMATIC, WaterslidePonderScene::ghostBlocks)
        registry.addStoryBoard(track, SLIDE_SECTOR_SCHEMATIC, WaterslidePonderScene::ghostBlocks)
        for (type in SlideAttachmentTypes.all()) {
            val binding = BuiltInRegistries.ITEM.getKey(type.item.get())
            registry.addStoryBoard(binding, ATTACHMENT_SCHEMATIC, WaterslidePonderScene::placeAttachment)
        }
        val door = BuiltInRegistries.ITEM.getKey(ModSlideAttachments.MECHANICAL_DOOR.item.get())
        registry.addStoryBoard(door, ATTACHMENT_SCHEMATIC, WaterslidePonderScene::mechanicalDoor)
        val detector = BuiltInRegistries.ITEM.getKey(ModSlideAttachments.SLIDE_DETECTOR.item.get())
        registry.addStoryBoard(detector, ATTACHMENT_SCHEMATIC, WaterslidePonderScene::slideDetector)
    }
}
